six/modules/c++/six.sicd/source/Grid.cpp

```diff
@@ -84,4 +480,812 @@ bool Grid::operator==(const Grid& rhs) const
+}
+
+double Grid::derivedColKCenter(const RMAT& rmat, double fc) const
+{
+    const double kfc = fc * 2 / math::Constants::SPEED_OF_LIGHT_METERS_PER_SEC;
+    return kfc * std::cos(rmat.dopConeAngleRef *
+        math::Constants::DEGREES_TO_RADIANS);
+}
+
+Vector3 Grid::derivedRowUnitVector(const RMAT& rmat, const Vector3& scp) const
+{
+    return rmat.uXCT(scp);
+}
+
+Vector3 Grid::derivedColUnitVector(const RMAT& rmat, const Vector3& scp) const
+{
+    return rmat.uYAT(scp);
+}
+
+Vector3 Grid::derivedRowUnitVector(const RMCR& rmcr, const Vector3& scp) const
+{
+    return rmcr.uXRG(scp);
+}
+
+Vector3 Grid::derivedColUnitVector(const RMCR& rmcr, const Vector3& scp) const
+{
+    return rmcr.uYCR(scp);
+}
+
+Vector3 Grid::derivedRowUnitVector(const INCA& inca, const Vector3& scp,
+        const PolyXYZ& arpPoly) const
+{
+    return inca.uRG(scp, arpPoly);
+}
+
+Vector3 Grid::derivedColUnitVector(const INCA& inca, const Vector3& scp,
+    const PolyXYZ& arpPoly) const
+{
+    return inca.uAZ(scp, arpPoly);
+}
+
+void Grid::fillDefaultFields(const RMCR& rmcr, double fc)
+{
+    if (!Init::isUndefined<double>(fc))
+    {
+        if (Init::isUndefined<double>(row->kCenter))
+        {
+            row->kCenter = derivedRowKCenter(rmcr, fc);
+        }
+        if (Init::isUndefined<double>(col->kCenter))
+        {
+            col->kCenter = 0;
+        }
+    }
+}
+
+void Grid::fillDefaultFields(const PFA& pfa, double fc)
+{
+    if (type == ComplexImageGridType::NOT_SET)
+    {
+        type = ComplexImageGridType::RGAZIM;
+    }
+
+    if (Init::isUndefined<double>(col->kCenter))
+    {
+        col->kCenter = 0;
+    }
+    if (Init::isUndefined<double>(row->kCenter))
+    {
+        if (!Init::isUndefined<double>(pfa.krg1) &&
+            !Init::isUndefined<double>(pfa.krg2))
+        {
+            // Default: the most reasonable way to compute this
+            row->kCenter = (pfa.krg1 + pfa.krg2) / 2;
+        }
+        else if (!Init::isUndefined<double>(fc))
+        {
+            // Approximation: this may not be quite right, due to
+            // rectangular inscription loss in PFA, but it should
+            // be close.
+            row->kCenter = fc * 
+                (2 / math::Constants::SPEED_OF_LIGHT_METERS_PER_SEC) *
+                pfa.spatialFrequencyScaleFactorPoly[0];
+        }
+    }
+}
+
+bool Grid::validate(const RMA& rma, const Vector3& scp,
+        const PolyXYZ& arpPoly, double fc,
+        logging::Logger& log) const
+{
+    bool valid = true;
+    // 2.12.3.2.1, 2.12.3.4.1
+    if (type != defaultGridType(rma))
+    {
+        std::ostringstream messageBuilder;
+        messageBuilder << "Given image formation algorithm expects "
+            << defaultGridType(rma).toString() << ".\nFound " << type;
+        log.error(messageBuilder.str());
+        valid = false;
+    }
+
+    if (rma.rmat.get())
+    {
+        return validate(*rma.rmat, scp, fc, log);
+    }
+
```

Comment:
In chained `if / else if` spots like this, don't have a newline between the blocks